File: src/main/java/net/dreamerzero/ClientCatcher/listener/ModListener.java
package net.dreamerzero.clientcatcher.listener;

import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.player.PlayerModInfoEvent;
import com.velocitypowered.api.proxy.Player;

import net.dreamerzero.clientcatcher.ModdedClient;

public class ModListener {
    @Subscribe
    public void onModInfo(PlayerModInfoEvent event){
        final Player player = event.getPlayer();
        final ModdedClient mclient = ModdedClient.getModdedClient(player.getUniqueId());
        mclient.setMods(event.getModInfo());
    }
}

File: src/main/kotlin/me/adrianed/clientcatcher/velocity/listener/ModListener.kt
package me.adrianed.clientcatcher.velocity.listener

import com.velocitypowered.api.event.Continuation
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.player.PlayerModInfoEvent
import me.adrianed.clientcatcher.velocity.ClientCatcher
import me.adrianed.clientcatcher.velocity.asMiniMessage
import me.adrianed.clientcatcher.velocity.event.BlockedModEvent
import me.adrianed.clientcatcher.velocity.objects.CatcherCommandSource
import net.kyori.adventure.permission.PermissionChecker

class ModListener(private val plugin: ClientCatcher) {
    @Subscribe
    fun onBrand(event: PlayerModInfoEvent, continuation: Continuation) {
        plugin.proxyServer
            .filterAudience {
                it.get(PermissionChecker.POINTER).map {
                        pointer -> pointer.test("clientcatcher.alert.mods")
                }.orElse(false)
            }.sendMessage(plugin.messages.alert.mods.asMiniMessage())

        for (mod in event.modInfo.mods) {
            for (blocked in plugin.configuration.blocked.mods) {
                if (blocked.key == mod.id) {
                    for (command in blocked.value) {
                        plugin.commandManager.executeAsync(CatcherCommandSource, command)
                        plugin.eventManager.fireAndForget(BlockedModEvent(mod, event.player))
                        continuation.resume()
                        return
                    }
                }
            }
        }
        continuation.resume()
    }
}